runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -1991,23 +1990,9 @@ MM_CopyForwardScheme::copy(MM_EnvironmentVLHGC *env, MM_AllocationContextTarok *
 
 				if (objectModel->isIndexable(destinationObjectPtr)) {
 					_extensions->indexableObjectModel.fixupInternalLeafPointersAfterCopy((J9IndexableObject *)destinationObjectPtr, (J9IndexableObject *)forwardedHeader->getObject());
-
-					/* Updates internal data address of indexable objects. Every indexable object have a void *dataAddr
-					 * that always points to the array data. It will always point to the address right after the header,
-					 * in case of contiguous data it will point to the data itself, and in case of discontiguous
-					 * arraylet it will point to the first arrayiod. dataAddr is only updated if dataAddr points to data
-					 * within heap. */
-					_extensions->indexableObjectModel.fixupDataAddr(destinationObjectPtr);
-				}
-
-				/* IF the object has been hashed and has not been moved then we must store the previous
-				 * address into the hashcode slot at hashcode offset.
-				 */
-				if (doesObjectNeedHash) {
-					UDATA hashOffset = objectModel->getHashcodeOffset(destinationObjectPtr);
-					U_32 *hashCodePointer = (U_32*)((U_8*)destinationObjectPtr + hashOffset);
-					*hashCodePointer = objectModel->computeObjectHash(forwardedHeader);
-					objectModel->setObjectHasBeenMoved(destinationObjectPtr);
+					_extensions->objectModel.fixupForwardedObjectIndexable(forwardedHeader, destinationObjectPtr, 0);
+				} else {
+					_extensions->objectModel.fixupForwardedObjectMixed(forwardedHeader, destinationObjectPtr, 0);
```

Comment:
Common objectModel is already cached into a local (no need for '_extensions->')

But as I stated before, you might consider calling specific object models directly (don't know if that's better or not)

_extensions->indexableObjectModel.fixupForwardedObject()
_extensions->mixedObjectModel.fixupForwardedObject()

Reply:
Ok, would this be the leading solution to investigate?

Reply:
Once we introduce a helper just for Hashing fixup this block would keep its original form and we would not waste time setting age bits:

isIndexable
    fixupLeafs
    fixupDataAddr

fixupHashing

Reply:
.... which overall means that I'm giving up on specialized mixed/indexable APIs after all (I did not see much benefit for them in Scavenger either).

Reply:
Either I was not clear enough, or you did not notice my comments, but I wanted you to restore this code to what it was (except calling the new common hashing helper instead of specialized CopyForward code block just below).

Reply:
Ok 👍  I thought what I implemented was consistent with your comments in https://github.com/eclipse-openj9/openj9/pull/12661#discussion_r634493167, but it seems like I have misunderstood. I will restore this code to what it was.

Reply:
Also, I have slightly simplified the hot field alignment code and also make it resemble more to Scavenger hot field alignment code. I hope it is fine to do that within this PR.